File: src/dots/commands/link.py
import typer
import shutil
from pathlib import Path
from rich.tree import Tree
from dots.ui.output import console, print_header, print_warning, print_error, print_info
from dots.core.config import DotsConfig
from dots.core.resolver import (
    resolve_modules,
    get_module_variant_info,
    get_module_available_sources,
    get_active_variant,
)
from dots.core.yaml_parser import detect_variants, parse_path_yaml
from dots.core.transaction import TransactionLog
from dots.ui.selector import select_modules


def _get_effective_variant(config, module_name: str, variant: str | None) -> str | None:
    """Returns the variant that will be used — explicit or cascade default."""
    if variant:
        return variant
    yaml_path = config.repo_root / module_name / "path.yaml"
    if not yaml_path.exists():
        return None
    mappings = parse_path_yaml(yaml_path, config.current_os)
    vinfo = detect_variants(mappings)
    if vinfo.has_variants:
        return vinfo.default_variant
    return None


def link_cmd(
    module: list[str] | None = typer.Option(
        None,
        "--module",
        "-m",
        help="Link only specific modules (repeatable: -m Zsh -m Nvim)",
    ),
    type: list[str] | None = typer.Option(
        None,
        "--type",
        "-t",
        help="Link only modules of this type (repeatable: -t minimal -t work)",
    ),
    dry_run: bool = typer.Option(False, "--dry-run", help="Show what would happen"),
    force: bool = typer.Option(False, "--force", help="Overwrite existing symlinks"),
    interactive: bool = typer.Option(
        False, "--interactive", "-i", help="Interactively select modules to link"
    ),
    variant: str | None = typer.Option(
        None,
        "--variant",
        help="Specific variant to use (when module has multiple variants sharing same destination)",
    ),
):
    """
    Create symlinks for dotfiles modules.

    By default, links all modules. Use --interactive to select specific modules.
    """
    print_header("Linking Dotfiles")

    config = DotsConfig.load()

    # Handle module selection
    if module:
        selected_modules = module
        print_info(f"Linking specified modules: {', '.join(selected_modules)}")
    elif interactive:
        selected_modules = select_modules(config, preselect_all=True)
        if not selected_modules:
            print_info("No modules selected.")
            return
        print_info(f"Selected {len(selected_modules)} module(s)")
        console.print()
    else:
        selected_modules = None  # Link all modules

    # If variant is specified, module must also be specified
    if variant and not selected_modules:
        print_error("When using --variant, you must specify the module name.")
        print_info("Example: dots link -m Nvim --variant notevim")
        raise typer.Exit(1)

    # When --variant is explicitly requested, auto-swap active variant conflicts.
    # If the destination already has a symlink to a different variant,
    # we treat it as a requested force-overwrite (variant switch).
    is_variant_swap = False
    if variant and not force:
        for module_name in selected_modules or []:
            active = get_active_variant(config, module_name)
            if active and active != variant:
                force = True
                is_variant_swap = True
                print_info(f"Auto-swap: {module_name} variant '{active}' → '{variant}'")
                break

    # Validate variant if specified with specific modules
    if variant and selected_modules:
        for module_name in selected_modules:
            variant_info = get_module_variant_info(config, module_name)
            available_sources = get_module_available_sources(config, module_name)

            if not variant_info or not variant_info.has_variants:
                # User specified a Variant but module has no variants
                print_error(
                    f"Module '{module_name}' has no variants. "
                    f"Available sources: {', '.join(available_sources)}"
                )
                raise typer.Exit(1)

            if variant not in variant_info.variants:
                print_error(
                    f"Variant '{variant}' not found in module '{module_name}'. "
                    f"Available variants: {', '.join(variant_info.variants)}"
                )
                raise typer.Exit(1)

    # Show warning when using cascade (variant auto-selected without explicit --variant)
    if not variant and selected_modules:
        for module_name in selected_modules:
            variant_info = get_module_variant_info(config, module_name)
            if variant_info and variant_info.has_variants:
                print_warning(
                    f"Module '{module_name}' has multiple variants. "
                    f"Using default: '{variant_info.default_variant}' (last in YAML). "
                    f"Use --variant to select a specific one."
                )

    # Pass variant to resolver (cascade handled inside if variant is None)
    type_filter = type if isinstance(type, list) else None
    modules = resolve_modules(
        config, modules=selected_modules, types=type_filter, variant=variant
    )

    if not modules:
        print_warning("No modules found.")
        return

    stats = {"linked": 0, "conflicts": 0, "pending": 0, "errors": 0}
    transaction = TransactionLog()

    try:
        for module_name, statuses in modules.items():
            # Create tree for this module
            module_tree = Tree(f"📦 [bold cyan]{module_name}[/bold cyan]")
            module_stats = {"linked": 0, "conflicts": 0, "pending": 0}

            # Calculate effective variant for this module
            effective_variant = _get_effective_variant(config, module_name, variant)
            variant_tag = (
                f" [dim][{effective_variant}][/dim]" if effective_variant else ""
            )

            for status in statuses:
                src = status.source
                final_dest = status.destination

                # Display based on state
                if status.state == "linked":
                    module_tree.add(f"[green]✔[/green] {src.name} → {final_dest}")
                    module_stats["linked"] += 1

                elif status.state == "unsafe":
                    module_tree.add(
                        f"[red]✗[/red] {src.name} → {final_dest} [red]({status.detail})[/red]"
                    )
                    module_stats["conflicts"] += 1

                elif status.state == "conflict":
                    if force:
                        if dry_run:
                            if is_variant_swap:
                                active = get_active_variant(config, module_name)
                                module_tree.add(
                                    f"[cyan]↔[/cyan] {src.name} → {final_dest} "
                                    f"[cyan](swapped: {active} → {variant})[/cyan]"
                                )
                            else:
                                module_tree.add(
                                    f"[yellow]⚠[/yellow] {src.name} → {final_dest} "
                                    f"[yellow](to be overwritten)[/yellow]"
                                )
                            module_stats["pending"] += 1
                        else:
                            if is_variant_swap:
                                active = get_active_variant(config, module_name)
                                module_tree.add(
                                    f"[cyan]↔[/cyan] {src.name} → {final_dest} "
                                    f"[cyan](swapped: {active} → {variant})[/cyan]"
                                )
                            else:
                                module_tree.add(
                                    f"[green]✔[/green] {src.name} → {final_dest} "
                                    f"[green](overwritten)[/green]"
                                )
                            module_stats["linked"] += 1
                            transaction.unlink(final_dest)
                            transaction.symlink(final_dest, src.resolve())
                    else:
                        module_tree.add(
                            f"[red]⚠[/red] {src.name} → {final_dest} [red](conflict: {status.detail})[/red]"
                        )
                        module_stats["conflicts"] += 1

                elif status.state == "pending":
                    if status.detail == "backup needed":
                        backup_path = Path(str(final_dest) + "-backup")
                        if backup_path.exists():
                            # Backup already exists -> require user attention and do not auto-create another backup
                            module_tree.add(
                                f"[yellow]⚠[/yellow] {src.name} → {final_dest} [yellow](backup exists at {backup_path}, review before creating new backups)[/yellow]"
                            )
                            module_stats["pending"] += 1
                        else:
                            if dry_run:
                                module_tree.add(
                                    f"[yellow]⚠[/yellow] {src.name} → {final_dest} [yellow](backup needed)[/yellow]"
                                )
                                module_stats["pending"] += 1
                            else:
                                module_tree.add(
                                    f"[green]✔[/green] {src.name} → {final_dest} [green](backed up and created)[/green]{variant_tag}"
                                )
                                module_stats["linked"] += 1
                                transaction.backup(final_dest, backup_path)
                                transaction.symlink(final_dest, src.resolve())
                    else:
                        if dry_run:
                            module_tree.add(
                                f"[blue]ℹ[/blue] {src.name} → {final_dest} [dim](to be created)[/dim]"
                            )
                            module_stats["pending"] += 1
                        else:
                            module_tree.add(
                                f"[green]✔[/green] {src.name} → {final_dest} [green](created)[/green]{variant_tag}"
                            )
                            module_stats["linked"] += 1
                            if not final_dest.parent.exists():
                                transaction.mkdir(final_dest.parent)
                            transaction.symlink(final_dest, src.resolve())

            # Add status line
            status_parts = []
            if module_stats["linked"] > 0:
                status_parts.append(f"[green]{module_stats['linked']} linked[/green]")
            if dry_run and module_stats["pending"] > 0:
                status_parts.append(
                    f"[yellow]{module_stats['pending']} to link[/yellow]"
                )
            if module_stats["conflicts"] > 0:
                status_parts.append(f"[red]{module_stats['conflicts']} conflicts[/red]")

            if status_parts:
                module_tree.add(f"[dim]Status: {' • '.join(status_parts)}[/dim]")

            # Update global stats
            for key in stats:
                stats[key] += module_stats.get(key, 0)

            console.print(module_tree)
            console.print()

        # Commit transaction if successful
        if not dry_run:
            transaction.commit()

    except Exception as e:
        # Rollback on any error
        if not dry_run:
            print_error(f"Error during linking: {e}")
            console.print("[yellow]Rolling back changes...[/yellow]")
            transaction.rollback()
            console.print("[green]Rollback complete.[/green]")
        stats["errors"] += 1
        raise typer.Exit(1)

    # Summary with divider
    console.print("━" * console.width)
    summary_parts = []
    if stats["linked"] > 0:
        summary_parts.append(f"[green]{stats['linked']} ✔ linked[/green]")
    if stats["conflicts"] > 0:
        summary_parts.append(f"[red]{stats['conflicts']} ⚠ conflicts[/red]")
    if dry_run and stats["pending"] > 0:
        summary_parts.append(f"[yellow]{stats['pending']} ℹ to link[/yellow]")

    console.print(f"[bold]Summary:[/bold] {' • '.join(summary_parts)}")

    if dry_run:
        console.print("\n[dim]This was a dry run. No changes were made.[/dim]")

File: internal/cli/link.go
package cli

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/Wilberucx/dots/internal/config"
	"github.com/Wilberucx/dots/internal/plan"
	"github.com/Wilberucx/dots/internal/resolver"
	"github.com/Wilberucx/dots/internal/transaction"
	"github.com/Wilberucx/dots/internal/ui"
	"github.com/spf13/cobra"
)

func init() {
	linkCmd.RunE = func(cmd *cobra.Command, args []string) error {
		return runLink(cmd)
	}
}

// linkRow is a display row for one mapping.
type linkRow struct {
	icon    string
	src     string
	dest    string
	detail  string
	variant string
}

// stateCount tracks link operation statistics per module.
type stateCount struct {
	linked    int
	conflicts int
	pending   int
	skipped   int
}

// linkTx wraps TransactionLog with error tracking for rollback.
type linkTx struct {
	tx   *transaction.TransactionLog
	fail error
}

func (l *linkTx) symlink(path, target string) {
	if l.fail != nil {
		return
	}
	if err := l.tx.Symlink(path, target); err != nil {
		l.fail = err
	}
}

func (l *linkTx) unlink(path string) {
	if l.fail != nil {
		return
	}
	if err := l.tx.Unlink(path); err != nil {
		l.fail = err
	}
}

func (l *linkTx) mkdir(path string) {
	if l.fail != nil {
		return
	}
	if err := l.tx.Mkdir(path); err != nil {
		l.fail = err
	}
}

func (l *linkTx) backup(path, backupPath string) {
	if l.fail != nil {
		return
	}
	if err := l.tx.Backup(path, backupPath); err != nil {
		l.fail = err
	}
}

func (l *linkTx) commit() error {
	if l.fail != nil {
		l.tx.Rollback()
		return l.fail
	}
	l.tx.Commit()
	return nil
}

func runLink(cmd *cobra.Command) error {
	ui.PrintHeader("Linking Dotfiles")

	cfg, err := loadConfig()
	if err != nil {
		return fmt.Errorf("loading config: %w", err)
	}

	modules := stringSliceFlag(cmd, "module")
	types := stringSliceFlag(cmd, "type")
	dryRun := boolFlag(cmd, "dry-run")
	force := boolFlag(cmd, "force")
	interactive := boolFlag(cmd, "interactive")
	variant := stringFlag(cmd, "variant")

	// ── Module selection ──────────────────────────────────────────────
	var selectedModules []string
	if len(modules) > 0 {
		selectedModules = modules
		ui.PrintInfo(fmt.Sprintf("Linking specified modules: %s", strings.Join(selectedModules, ", ")))
	} else if interactive {
		selectedModules = selectModulesInteractive(cfg, true)
		if len(selectedModules) == 0 {
			ui.PrintInfo("No modules selected.")
			return nil
		}
		ui.PrintInfo(fmt.Sprintf("Selected %d module(s)", len(selectedModules)))
		fmt.Println()
	} else {
		selectedModules = nil // Link all modules
	}

	// ── Variant validation ────────────────────────────────────────────
	if variant != "" && len(selectedModules) == 0 {
		ui.PrintError("When using --variant, you must specify the module name.")
		ui.PrintInfo("Example: dots link -m Nvim --variant notevim")
		return fmt.Errorf("--variant requires --module")
	}

	// Variant auto-swap: track which modules need a swap per-module
	variantSwapModules := buildVariantSwapMap(cfg, selectedModules, variant, force)

	// Print auto-swap messages (caller's responsibility, not buildVariantSwapMap's)
	for _, modName := range sortedSwapModules(variantSwapModules) {
		active, _ := resolver.GetActiveVariant(cfg, modName)
		ui.PrintInfo(fmt.Sprintf("Auto-swap: %s variant '%s' → '%s'", modName, active, variant))
	}

	// Validate variant existence
	if variant != "" && len(selectedModules) > 0 {
		for _, modName := range selectedModules {
			vInfo, err := resolver.GetModuleVariantInfo(cfg, modName)
			if err != nil {
				return fmt.Errorf("checking variant info for %s: %w", modName, err)
			}
			if vInfo == nil || !vInfo.HasVariants {
				modDir := filepath.Join(cfg.RepoRoot, modName)
				entries, _ := os.ReadDir(modDir)
				var sources []string
				for _, e := range entries {
					if !e.IsDir() && e.Name() != "path.yaml" {
						sources = append(sources, e.Name())
					} else if e.IsDir() {
						sources = append(sources, e.Name()+"/")
					}
				}
				return fmt.Errorf("module '%s' has no variants (available: %s)", modName, strings.Join(sources, ", "))
			}
			found := false
			for _, v := range vInfo.Variants {
				if v == variant {
					found = true
					break
				}
			}
			if !found {
				return fmt.Errorf("variant '%s' not found in module '%s' (available: %s)", variant, modName, strings.Join(vInfo.Variants, ", "))
			}
		}
	}

	// Cascade warning
	if variant == "" && len(selectedModules) > 0 {
		for _, modName := range selectedModules {
			vInfo, _ := resolver.GetModuleVariantInfo(cfg, modName)
			if vInfo != nil && vInfo.HasVariants {
				ui.PrintWarning(fmt.Sprintf(
					"Module '%s' has multiple variants. Using default: '%s'. Use --variant to select a specific one.",
					modName, vInfo.DefaultVariant,
				))
			}
		}
	}

	// ── Resolve and build plan ────────────────────────────────────────
	allModules, err := resolver.ResolveModules(cfg, selectedModules, types, variant)
	if err != nil {
		return fmt.Errorf("resolving modules: %w", err)
	}
	if len(allModules) == 0 {
		ui.PrintWarning("No modules found.")
		return nil
	}

	// Determine effective variant for display
	getEffectiveVariant := func(moduleName string) string {
		if variant != "" {
			return variant
		}
		vInfo, err := resolver.GetModuleVariantInfo(cfg, moduleName)
		if err == nil && vInfo != nil && vInfo.HasVariants {
			return vInfo.DefaultVariant
		}
		return ""
	}

	// Build the plan — centralized decision logic
	planOpts := plan.BuildOptions{
		Force:        force,
		VariantSwaps: variantSwapModules,
	}
	p := plan.BuildLinkPlan(allModules, planOpts)

	// ── Process plan actions ──────────────────────────────────────────
	stats := &stateCount{}
	tx := &linkTx{tx: &transaction.TransactionLog{}}
	byModule := p.ActionsByModule()

	for _, modName := range sortedMapKeys(byModule) {
		actions := byModule[modName]
		modStats := &stateCount{}
		var rows []linkRow

		// Effective variant for this module
		effVariant := getEffectiveVariant(modName)
		variantTag := ""
		if effVariant != "" {
			variantTag = fmt.Sprintf(" [%s]", effVariant)
		}

		for _, a := range actions {
			if tx.fail != nil {
				break
			}
			srcName := filepath.Base(a.Source)
			shortDest := shortDisplayPath(a.Destination, cfg.HomeDir)

			switch a.Kind {
			case plan.ActionCreateSymlink:
				if dryRun {
					rows = append(rows, linkRow{
						icon:   ui.InfoStyle.Render(ui.IconPending),
						src:    srcName,
						dest:   shortDest,
						detail: ui.DimStyle.Render("(to be created)"),
					})
					modStats.pending++
				} else {
					rows = append(rows, linkRow{
						icon:   ui.SuccessStyle.Render(ui.IconLinked),
						src:    srcName,
						dest:   shortDest,
						detail: ui.SuccessStyle.Render(fmt.Sprintf("(created)%s", variantTag)),
					})
					modStats.linked++
					parentDir := filepath.Dir(a.Destination)
					if _, err := os.Stat(parentDir); os.IsNotExist(err) {
						tx.mkdir(parentDir)
					}
					tx.symlink(a.Destination, a.Source)
				}

			case plan.ActionBackupFile:
				backupPath := a.BackupPath
				if backupPath == "" {
					backupPath = a.Destination + ".orig"
				}
				if dryRun {
					rows = append(rows, linkRow{
						icon:   ui.WarningStyle.Render(ui.IconConflict),
						src:    srcName,
						dest:   shortDest,
						detail: ui.WarningStyle.Render("(.orig needed)"),
					})
					modStats.pending++
				} else {
					rows = append(rows, linkRow{
						icon:   ui.SuccessStyle.Render(ui.IconLinked),
						src:    srcName,
						dest:   shortDest,
						detail: ui.SuccessStyle.Render(fmt.Sprintf("(.orig saved and linked)%s", variantTag)),
					})
					modStats.linked++
					parentDir := filepath.Dir(a.Destination)
					if _, err := os.Stat(parentDir); os.IsNotExist(err) {
						tx.mkdir(parentDir)
					}
					tx.backup(a.Destination, backupPath)
					tx.symlink(a.Destination, a.Source)
				}

			case plan.ActionReplaceSymlink:
				moduleIsSwap := variantSwapModules[modName]
				if dryRun {
					if moduleIsSwap {
						active, _ := resolver.GetActiveVariant(cfg, modName)
						rows = append(rows, linkRow{
							icon:   ui.InfoStyle.Render(ui.IconSwap),
							src:    srcName,
							dest:   shortDest,
							detail: ui.InfoStyle.Render(fmt.Sprintf("(swapped: %s → %s)", active, variant)),
						})
					} else {
						rows = append(rows, linkRow{
							icon:   ui.WarningStyle.Render(ui.IconConflict),
							src:    srcName,
							dest:   shortDest,
							detail: ui.WarningStyle.Render("(to be overwritten)"),
						})
					}
					modStats.pending++
				} else {
					if moduleIsSwap {
						active, _ := resolver.GetActiveVariant(cfg, modName)
						rows = append(rows, linkRow{
							icon:   ui.InfoStyle.Render(ui.IconSwap),
							src:    srcName,
							dest:   shortDest,
							detail: ui.InfoStyle.Render(fmt.Sprintf("(swapped: %s → %s)", active, variant)),
						})
					} else {
						rows = append(rows, linkRow{
							icon:   ui.SuccessStyle.Render(ui.IconLinked),
							src:    srcName,
							dest:   shortDest,
							detail: ui.SuccessStyle.Render("(overwritten)"),
						})
					}
					modStats.linked++
					parentDir := filepath.Dir(a.Destination)
					if _, err := os.Stat(parentDir); os.IsNotExist(err) {
						tx.mkdir(parentDir)
					}
					tx.unlink(a.Destination)
					tx.symlink(a.Destination, a.Source)
				}

			case plan.ActionSkipLinked:
				rows = append(rows, linkRow{
					icon:   ui.SuccessStyle.Render(ui.IconLinked),
					src:    srcName,
					dest:   shortDest,
					detail: "",
				})
				modStats.linked++

			case plan.ActionSkipPending:
				rows = append(rows, linkRow{
					icon:   ui.DimStyle.Render("○"),
					src:    srcName,
					dest:   shortDest,
					detail: ui.DimStyle.Render(fmt.Sprintf("(%s)", a.Detail)),
				})
				modStats.skipped++

			case plan.ActionErrorConflict:
				rows = append(rows, linkRow{
					icon:   ui.ErrorStyle.Render(ui.IconConflict),
					src:    srcName,
					dest:   shortDest,
					detail: ui.ErrorStyle.Render(fmt.Sprintf("(conflict: %s)", a.Detail)),
				})
				modStats.conflicts++

			case plan.ActionErrorUnsafe:
				rows = append(rows, linkRow{
					icon:   ui.ErrorStyle.Render(ui.IconError),
					src:    srcName,
					dest:   shortDest,
					detail: ui.ErrorStyle.Render(fmt.Sprintf("(%s)", a.Detail)),
				})
				modStats.conflicts++
			}
		}

		// Print module tree
		fmt.Printf("  %s %s\n", ui.DimStyle.Render(ui.IconModule), ui.BoldStyle.Render(modName))
		for _, row := range rows {
			msg := fmt.Sprintf("    %s %s → %s", row.icon, row.src, row.dest)
			if row.detail != "" {
				msg += " " + row.detail
			}
			fmt.Println(msg)
		}

		// Module status line
		var statusParts []string
		if modStats.linked > 0 {
			statusParts = append(statusParts, ui.SuccessStyle.Render(fmt.Sprintf("%d linked", modStats.linked)))
		}
		if dryRun && modStats.pending > 0 {
			statusParts = append(statusParts, ui.WarningStyle.Render(fmt.Sprintf("%d to link", modStats.pending)))
		}
		if modStats.conflicts > 0 {
			statusParts = append(statusParts, ui.ErrorStyle.Render(fmt.Sprintf("%d conflicts", modStats.conflicts)))
		}
		if modStats.skipped > 0 {
			statusParts = append(statusParts, ui.DimStyle.Render(fmt.Sprintf("%d skipped", modStats.skipped)))
		}
		if len(statusParts) > 0 {
			fmt.Printf("    %s\n", ui.DimStyle.Render(fmt.Sprintf("Status: %s", strings.Join(statusParts, " • "))))
		}
		fmt.Println()

		stats.linked += modStats.linked
		stats.conflicts += modStats.conflicts
		stats.pending += modStats.pending
		stats.skipped += modStats.skipped
	}

	// ── Commit or rollback ────────────────────────────────────────────
	if !dryRun {
		if err := tx.commit(); err != nil {
			ui.PrintError(fmt.Sprintf("Error during linking: %v", err))
			ui.PrintInfo("Rolling back changes...")
			ui.PrintSuccess("Rollback complete.")
			return fmt.Errorf("linking failed: %w", err)
		}
	}

	// ── Summary ───────────────────────────────────────────────────────
	ui.PrintDivider(0)
	summaryParts := []string{}
	if stats.linked > 0 {
		summaryParts = append(summaryParts, ui.SuccessStyle.Render(fmt.Sprintf("%d ✔ linked", stats.linked)))
	}
	if stats.conflicts > 0 {
		summaryParts = append(summaryParts, ui.ErrorStyle.Render(fmt.Sprintf("%d ⚠ conflicts", stats.conflicts)))
	}
	if dryRun && stats.pending > 0 {
		summaryParts = append(summaryParts, ui.WarningStyle.Render(fmt.Sprintf("%d ℹ to link", stats.pending)))
	}
	if stats.skipped > 0 {
		summaryParts = append(summaryParts, ui.DimStyle.Render(fmt.Sprintf("%d skipped", stats.skipped)))
	}
	fmt.Printf("%s %s\n", ui.BoldStyle.Render("Summary:"), strings.Join(summaryParts, " • "))

	if dryRun {
		fmt.Printf("\n%s\n", ui.DimStyle.Render("This was a dry run. No changes were made."))
	}

	return nil
}

// sortedSwapModules returns sorted keys from a variant swap map for deterministic output.
func sortedSwapModules(m map[string]bool) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

// buildVariantSwapMap detects which selected modules need a variant swap.
// A module needs a swap when:
//   - --variant is specified
//   - --force is NOT set
//   - the module has an active variant different from the requested one
//
// Returns nil (no swaps) when conditions aren't met.
// This function is PURE — it does NOT print anything. Callers are responsible
// for rendering any UI messages.
func buildVariantSwapMap(cfg *config.DotsConfig, selectedModules []string, variant string, force bool) map[string]bool {
	if variant == "" || force || len(selectedModules) == 0 {
		return nil
	}

	swapMap := make(map[string]bool)
	for _, modName := range selectedModules {
		active, _ := resolver.GetActiveVariant(cfg, modName)
		if active != "" && active != variant {
			swapMap[modName] = true
		}
	}

	if len(swapMap) == 0 {
		return nil
	}
	return swapMap
}

// sortedMapKeys returns sorted keys from a map of string to []plan.Action.
func sortedMapKeys(m map[string][]plan.Action) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

// selectModulesInteractive runs a Bubbletea TUI for interactive module selection.
func selectModulesInteractive(cfg *config.DotsConfig, preselectAll bool) []string {
	modDirs, err := cfg.GetModuleDirs(nil, nil)
	if err != nil || len(modDirs) == 0 {
		return nil
	}

	names := make([]string, len(modDirs))
	for i, d := range modDirs {
		names[i] = d.Name
	}

	selected, err := ui.RunModuleSelector(names, preselectAll)
	if err != nil {
		return nil
	}
	return selected
}
